New auth best practices doc (#44284)
dotnet/docs: docs/azure/sdk/authentication/authentication-best-practices.md

---
title: Authentication best practices with the Azure Identity library for .NET

> description: This article describes authentication best practices to follow when using the Azure Identity library.

Reviewer: ```suggestion
description: This article describes authentication best practices to follow when using the Azure Identity library for .NET.
```

ms.topic: conceptual
ms.date: 01/15/2025
---

# Authentication best practices with the Azure Identity library for .NET

This article offers guidelines to help you maximize the performance and reliability of your .NET apps when authenticating to Azure services. To make the most of the Azure Identity library for .NET, it's important to understand potential issues and mitigation techniques.

## Use deterministic credentials in production environments

[!INCLUDE [default-azure-credential-usage](../includes/default-azure-credential-usage.md)]

## Reuse credential instances

Reuse credential instances when possible to improve app resilience and reduce the number of access token requests issued to Microsoft Entra ID. When a credential is reused, an attempt is made to fetch a token from the app token cache managed by the underlying MSAL dependency. For more information, see [Token caching in the Azure Identity client library](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/samples/TokenCache.md).

> [!NOTE]
> A high-volume app that doesn't reuse credentials may encounter HTTP 429 throttling responses from Microsoft Entra ID, which can lead to app outages.

> In an ASP.NET Core app, implement credential reuse through the `UseCredential` method of `Microsoft.Extensions.Azure`:

Reviewer: ```suggestion
In an ASP.NET Core project, implement credential reuse through the `UseCredential` method of `Microsoft.Extensions.Azure`:
```

:::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_credential_reuse_Dac" highlight="6" :::

For information on this approach, see [Authenticate using Microsoft Entra ID](/dotnet/azure/sdk/aspnetcore-guidance?tabs=api#authenticate-using-microsoft-entra-id).

Other types of .NET apps can reuse credential instances as follows:

> :::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_credential_reuse_noDac" highlight="8, 12" :::

Reviewer: We should tidy this up by introducing 2 tabs: "ASP.NET Core" and "Other". That change reduces the vertical scrolling.

## Understand the managed identity retry strategy

The Azure Identity library for .NET allows you to authenticate via managed identity with `ManagedIdentityCredential`. The way you use `ManagedIdentityCredential` impacts the applied retry strategy:

- When used via `DefaultAzureCredential`:

>   - No retries are attempted when token acquisition fails, which makes this the least resilient option.

Reviewer 2: ```suggestion
  - No retries are attempted when the initial token acquisition attempt fails or times out after a short duration, which makes this the least resilient option.
```

- When used via any other approach, such as `ChainedTokenCredential` or `ManagedIdentityCredential` directly:

>   - The time interval between retries starts at 0.8 seconds, and a maximum of five retries are attempted.

Reviewer 2: ```suggestion
  - The time interval between retries starts at 0.8 seconds and increases exponentially. A maximum of five retries are attempted.
```

  - If the Azure service to which you're authenticating provides a `Retry-After` response header, the next retry is delayed by the duration specified in that header's value.

>   - If the service doesn't provide a `Retry-After` header, the maximum permissible delay between retries is 1 minute.

Reviewer 2: I would remove them.

> The Azure Identity library for .NET allows you to authenticate via managed identity with `ManagedIdentityCredential`. The way you use `ManagedIdentityCredential` impacts the applied retry strategy:
> 
> - When used via `DefaultAzureCredential`:
>   - No retries are attempted when the initial token acquisition attempt fails or times out after a short duration, which makes this the least resilient option.
> - When used via any other approach, such as `ChainedTokenCredential` or `ManagedIdentityCredential` directly:
>   - The time interval between retries starts at 0.8 seconds, and a maximum of five retries are attempted.
>   - To change any of the default retry settings, use the `Retry` property on `ManagedIdentityCredentialOptions`. For example, retry a maximum of three times, with a starting interval of 0.5 seconds:
> 
> :::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_retries" highlight="5-9" :::

Reviewer: ```suggestion
The Azure Identity library for .NET allows you to authenticate via managed identity with `ManagedIdentityCredential`. The way in which you use `ManagedIdentityCredential` impacts the applied retry strategy. When used via:

- `DefaultAzureCredential`, no retries are attempted when the initial token acquisition attempt fails or times out after a short duration, which makes this the least resilient option.
- Any other approach, such as `ChainedTokenCredential` or `ManagedIdentityCredential` directly:
  - The time interval between retries starts at 0.8 seconds, and a maximum of five retries are attempted.
  - To change any of the default retry settings, use the `Retry` property on `ManagedIdentityCredentialOptions`. For example, retry a maximum of three times, with a starting interval of 0.5 seconds:

    :::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_retries" highlight="5-9" :::
```

>   - To change any of the default retry settings, use the `Retry` property on `ManagedIdentityCredentialOptions`. For example, retry a maximum of three times, with a starting interval of 0.5 seconds:

Reviewer 2: Should we link to more information here? https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Configuration.md#setting-a-custom-retry-policy

> For more information on customizing retry policies, see [Setting a custom retry policy](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Configuration.md#setting-a-custom-retry-policy)

Reviewer: ```suggestion
For more information on customizing retry policies, see [Setting a custom retry policy](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Configuration.md#setting-a-custom-retry-policy).
```

---
title: Authentication best practices with the Azure Identity library for .NET
description: This article describes authentication best practices to follow when using the Azure Identity library for .NET.
ms.topic: conceptual
ms.date: 01/15/2025
---

# Authentication best practices with the Azure Identity library for .NET

This article offers guidelines to help you maximize the performance and reliability of your .NET apps when authenticating to Azure services. To make the most of the Azure Identity library for .NET, it's important to understand potential issues and mitigation techniques.

## Use deterministic credentials in production environments

> `DefaultAzureCredential` is the most approachable way to get started with the Azure Identity library, but that convenience also introduces certain tradeoffs. Most notably, the specific credential in the chain that will succeed and be used for request authentication can't be guaranteed ahead of time. In a production environment, this unpredictability can introduce significant and sometimes subtle problems.

Reviewer: Link this first occurrence of `DefaultAzureCredential` to the following section in the credential chains doc: https://learn.microsoft.com/dotnet/azure/sdk/authentication/credential-chains?tabs=dac#defaultazurecredential-overview

For example, consider the following `DefaultAzureCredential` configuration in an ASP.NET Core project:

:::code language="csharp" source="../snippets/authentication/credential-chains/Program.cs" id="snippet_Dac" highlight="6,7":::

Replace the preceding code with a `ChainedTokenCredential` implementation that specifies only the necessary credentials:

:::code language="csharp" source="../snippets/authentication/credential-chains/Program.cs" id="snippet_Ctc" highlight="6-8":::

In this example, `ManagedIdentityCredential` would be automatically discovered in production, while `VisualStudioCredential` would work in local development environments.

## Reuse credential instances

Reuse credential instances when possible to improve app resilience and reduce the number of access token requests issued to Microsoft Entra ID. When a credential is reused, an attempt is made to fetch a token from the app token cache managed by the underlying MSAL dependency. For more information, see [Token caching in the Azure Identity client library](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/samples/TokenCache.md).

> > [!NOTE]

Reviewer: Change the widget type to stress the importance of this message.
```suggestion
> [!IMPORTANT]
```

> A high-volume app that doesn't reuse credentials may encounter HTTP 429 throttling responses from Microsoft Entra ID, which can lead to app outages.

> Implement credential reuse through the `UseCredential` method of `Microsoft.Extensions.Azure`:
> 
> # [ASP.NET Core](#tab/aspdotnet)
> 
> :::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_credential_reuse_Dac" highlight="6" :::

Reviewer: ```suggestion
The recommended credential reuse strategy differs by .NET application type.

# [ASP.NET Core](#tab/aspdotnet)

Implement credential reuse through the `UseCredential` method of `Microsoft.Extensions.Azure`:

:::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_credential_reuse_Dac" highlight="6" :::
```

> [!IMPORTANT]
> A high-volume app that doesn't reuse credentials may encounter HTTP 429 throttling responses from Microsoft Entra ID, which can lead to app outages.

The recommended credential reuse strategy differs by .NET application type.

# [ASP.NET Core](#tab/aspdotnet)

Implement credential reuse through the `UseCredential` method of `Microsoft.Extensions.Azure`:

:::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_credential_reuse_Dac" highlight="6" :::

For information on this approach, see [Authenticate using Microsoft Entra ID](/dotnet/azure/sdk/aspnetcore-guidance?tabs=api#authenticate-using-microsoft-entra-id).

# [Other](#tab/other)

:::code language="csharp" source="../snippets/authentication/best-practices/Program.cs" id="snippet_credential_reuse_noDac" highlight="8, 12" :::

---

## Understand the managed identity retry strategy

The Azure Identity library for .NET allows you to authenticate via managed identity with `ManagedIdentityCredential`. The way in which you use `ManagedIdentityCredential` impacts the applied retry strategy. When used via:

> - `DefaultAzureCredential`, no retries are attempted when the initial token acquisition attempt fails or times out after a short duration, which makes this the least resilient option.

Reviewer 2: ```suggestion
- `DefaultAzureCredential`, no retries are attempted when the initial token acquisition attempt fails or times out after a short duration. This is the least resilient option because it is optimized to "fail fast" for an efficient development inner-loop.
```

- Any other approach, such as `ChainedTokenCredential` or `ManagedIdentityCredential` directly:

>   - The time interval between retries starts at 0.8 seconds, and a maximum of five retries are attempted.

Reviewer 2: ```suggestion
  - The time interval between retries starts at 0.8 seconds, and a maximum of five retries are attempted, by default. This option is optimized for resilience but introduces potentially unwanted delays in the development inner-loop.
```
